### PR TITLE
"ssl-legacy" and "ciphersuites" options for xe command

### PR DESCRIPTION
**Pull request by thomassa**
Based on the changes made for the xe produced by the xen-api repository:
xapi-project/xen-api#2587

### CODE REVIEW

File: src/channels.ml [hunk 1]
```diff
@@ -77,13 +77,47 @@ let of_seekable_fd fd =
...
+let default_good_ciphers = "!EXPORT:RSA+AES128-SHA256"
+let default_legacy_ciphers = "RSA+AES256-SHA:RSA+AES128-SHA:RSA+RC4-SHA:RSA+DES-CBC3-SHA"
+
+let sslctx_cached = ref None
+
+let gen_sslctx legacy ciphers =
+  let ctx =
+    (* Here SSLv23 means ALL protocols *)
+    Ssl.create_context
+      (if legacy then Ssl.SSLv23 else Ssl.TLSv1_2)
```
Comment (euanh) [inline src/channels.ml]: Why does SSLv23 mean all protocols when passed to Ssl.create_context, but not below where it is passed to Ssl.disable_protocols?   Would it be worthwhile adding a local alias called something like "SSL_all"?

File: src/channels.ml [hunk 2]
```diff
@@ -77,13 +77,47 @@ let of_seekable_fd fd =
...
+      Ssl.Client_context
+  in
+  if legacy then (
```
Comment (euanh) [inline src/channels.ml]: Should this be 'not legacy'?   It looks like we are disabling SSLv2 and 3 when we are in legacy mode, which seems the wrong way round to me.

File: src/channels.ml (hunk 1 above)
Comment (thomassa) [inline src/channels.ml]: Why, you ask? I don't know why, but this is how the Ssl module works: see
https://github.com/savonet/ocaml-ssl/blob/master/src/ssl.mli#L205
and https://github.com/savonet/ocaml-ssl/blob/master/src/ssl.mli#L285
SSLv23 means "all protocols, both SSL and TLS" as an argument to the `create_context` function, but means only SSL protocols when passed as an argument to the `disable_protocols` function.

I tried to make that clear through my comments saying what my code achieves, but perhaps I should have been even more explicit.

File: src/channels.ml (hunk 2 above)
Comment (thomassa) [inline src/channels.ml]: It's correct.
In legacy mode, we pass the "all protocols" value to the context-creation function, then disable the protocols we don't want.
In strict mode, we pass the "TLSv1.2 only" value to the context-creation function, so there is no need to disable any unwanted protocols afterwards.

I have tested this against a server with various server-side configurations, to make sure it connects using the intended protocols and no others.

Comment (euanh) [inline src/channels.ml]: I think you should add a comment saying more or less this.   Unless you already know the details of the SSL library, it looks like you just turn some stuff on and then immediately turn it off again.

**Comment by thomassa** [comment]
I've added to the comments.

**Comment by thomassa** [comment]
Merging this now that the Travis test has passed.

**Pull request merged**